### DISCUSSION TITLE
Export CSV files: Configurable separator

### DISCUSSION THREAD

**Discussion by frheinze** [directus/directus#12456] [Feature Requests]
I would like to have the possibility to configure the separator in the exported CSV files. The reason is that e.g. on German Windows operating systems Excel expects a ';' instead of the usual ','. What do you think?

**Comment by rijkvanzanten** [comment]
I like it! In order to achieve that, we'll have to add some sort of way to tell the API specific export options. We might have to add a new `export_options` query parameter, or potentially utilize the same export param, but differentiate between "flat" string and object use, for example:

```
?export=csv
&export_options[separator]=;
```

or

```
?export[format]=csv
&export[separator]=;
```

**Comment by dewita97** [comment]
Hi @rijkvanzanten, 

Following what Microsoft has implemented, it is possible to give a first line specifying the separator (take a look at [Open or import CSV with different delimiter](https://www.ablebits.com/office-addins-blog/change-excel-csv-delimiter/#opening)). 

In this case, no need adding extra `export_options` and it should also work for the import functionality in the App !

I think that the _default_ ',' separator is not suited for every countries. In fact, for Belgium, finding a way to save a csv file with ',' as a separator is a heavy pain... On MacOS, we need to change the language and location settings in order to export a file in the correct format, then we can fallback to the previous settings. 

That's not suitable for clients 😕

We should instead give the user/developer the possibility to choose/specify the delimiter. Is this reasonable ?

**Reply by rijkvanzanten** [reply]
Not sure if we should rely on a non/semi-standard first line in a CSV for that configuration though, it feels a little cleaner to me to handle that through a query parameter or similar, as that would allow you to use the separator config without having to update the original file 🤔

**Reply by dewita97** [reply]
Yep, I think you gave a better proposal !
In any case, it's an improvement to consider. I was searching a lot before finding a thread like this one talking about this "issue". In fact, I was expecting more people having this concern 🙄

**Comment by christianrr** [comment]
Another thought: Why not setting the CSV delimiter in the config file globally per instance? Since the CSV delimiter is mostly region-related (in Austria/Germany a semi-colon), it would most likely not change per project/instance and therefore the .env file would be IMO a good place to set.

**Reply by christianrr** [reply]
btw: this would be achievable with one line of code in the import/endpoint implementation:

```
importCSV(collection: string, stream: NodeJS.ReadableStream): Promise<void> {
...
stream
   .pipe(stripBomStream())
   .pipe(csv({ separator: env.CSV_DELIMITER }))
...
}
```

**Reply by rijkvanzanten** [reply]
> Since the CSV delimiter is mostly region-related (in Austria/Germany a semi-colon), it would most likely not change per project/instance and therefore the .env file would be IMO a good place to set.

That would lock out international users though 🙂 I also think the delimiter depends on the other software you're intending to use the exported file with. F.e. Apple's Numbers defaults to semi-colons in US region as well

**Reply by christianrr** [reply]
Agreed, but it would still be a decent quick shot IMO...

What we would ultimately need is some config option in the Admin interface, otherwise a query parameter is pretty useless for importing/exporting from the UI.

What about storing the CSV delimiter as sort of an override in the user profile? The import/export endpoints could then either use the global setting or the user-specific override. This would bring the international users back in 😉

**Reply by rijkvanzanten** [reply]
That could work! I'm also down to make it a dropdown during every export, and storing the last used delimiter in localstorage. That way you have the flexibility to do it on a case-by-case basis, but it defaults to your preferred method

**Reply by benhaynes** [reply]
Couldn't it be a config file option for now, that then becomes a default that can be overridden per export in the future? That way you can set it project-wide (default) or override if needed. Could still have the override stored in localstorage if we want.

**Reply by rijkvanzanten** [reply]
Technically speaking yes, but I'd rather prevent the tech overhead of implementing another env var default for something small that we know we're also resolving in the UI itself 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
